File: 24511/24511.cpp
/* ************************************************************************** */
/*                                                                            */
/*                                                        :::      ::::::::   */
/*   24511.cpp                                          :+:      :+:    :+:   */
/*                                                    +:+ +:+         +:+     */
/*   By: jihoolee <jihoolee@student.42SEOUL.kr>     +#+  +:+       +#+        */
/*                                                +#+#+#+#+#+   +#+           */
/*   Created: 2024/03/10 10:55:02 by jihoolee          #+#    #+#             */
/*   Updated: 2024/03/10 13:55:24 by jihoolee         ###   ########.fr       */
/*                                                                            */
/* ************************************************************************** */

#include <iostream>
#include <vector>
#include <deque>

void get_input(std::deque<int>* questack) {
  int len;
  int num;

  std::cin >> len;

  std::vector<int>  s_q(len);

  for (int idx = 0; idx < len; ++idx) {
    std::cin >> num;
    s_q[idx] = num;
  }
  for (int idx = 0; idx < len; ++idx) {
    std::cin >> num;
    if (s_q[idx] == 0)
      questack->push_front(num);
  }
}

int main(void) {
  std::ios::sync_with_stdio(false);
  std::cin.tie(nullptr);
  std::cout.tie(nullptr);

  std::deque<int> questack;
  int             input_cnt;
  int             num;

  get_input(&questack);
  std::cin >> input_cnt;
  for (int i = 0; i < input_cnt; ++i) {
    std::cin >> num;
    questack.push_back(num);
    std::cout << questack.front() << ' ';
    questack.pop_front();
  }
  return 0;
}

File: 24511/24511.py
# **************************************************************************** #
#                                                                              #
#                                                         :::      ::::::::    #
#    24511.py                                           :+:      :+:    :+:    #
#                                                     +:+ +:+         +:+      #
#    By: jihoolee <jihoolee@student.42SEOUL.kr>     +#+  +:+       +#+         #
#                                                 +#+#+#+#+#+   +#+            #
#    Created: 2024/03/10 11:00:20 by jihoolee          #+#    #+#              #
#    Updated: 2024/03/10 12:32:07 by jihoolee         ###   ########.fr        #
#                                                                              #
# **************************************************************************** #

import sys
from collections import deque

input = sys.stdin.readline


def main() -> None:
    _ = input()
    s_q: list[int] = list(map(int, input().split()))
    init_elem: list[int] = list(map(int, input().split()))
    _ = input()
    elems: list[int] = list(map(int, input().split()))

    qs: deque[int] = deque()
    for is_stack, elem in zip(s_q, init_elem):
        if not is_stack:
            qs.appendleft(elem)
    for elem in elems:
        qs.append(elem)
        print(qs.popleft(), end=' ')


if __name__ == '__main__':
    main()
